### PR TITLE
Clean up magic numbers for element classes.

### PR DESCRIPTION
**Pull request by Mr-Chi-Ding**
Summary:

(1) Use FEType class to replace magic numbers like 501, 502, 601, and 602.

(2) Rename FEAElement_Tet10_v2 and Quadpts_vis_Tet10_v2 as FEAElement_Tet10 and Quadpts_vis_Tet10.

(3) Remove some unused classes like Tet10_P2_P1 and NodalBC_3D_ring.

There still exist some magic numbers. See the attached figure.
截屏2024-12-23 21 36 12

### CODE REVIEW

**Comment by ju-liu** [comment]
@Mr-Chi-Ding 这里最终是要把全部程序里的magic number都替换成新的名字。

File: include/FEAElement_Tet4.hpp
Comment (ju-liu) [inline include/FEAElement_Tet4.hpp:1]: 改动：
分隔符为80个字符的长度（原来是70个）

get_Type函数更新

不再使用get_memory_usage函数（从来没有使用过，过度设计了。）

**Review by Syj-0525 — COMMENTED** [review]
编译错误，以前get_type返回int类型，现在类型变了，一些函数传入的参数类型应该需要修改：
       class QuadPts_on_face final : public IQuadPts
  {
    public:
      QuadPts_on_face(const int &vol_elemType, const int &face_id, 
          const IQuadPts * const lower_quad_rule);
          
      编译警告：
      |   PetscPrintf(PETSC_COMM_WORLD, "elemType: %d. \n", get_Type());
      |                                            ~^       ~~~~~~~~~~
      |                                             |               |
      |                                             int             FEType

File: examples/ns/src/PLocAssem_VMS_NS_GenAlpha.cpp [hunk 1]
```diff
@@ -5,17 +5,17 @@ PLocAssem_VMS_NS_GenAlpha::PLocAssem_VMS_NS_GenAlpha(
...
         const int &in_snlocbas,
         const double &in_rho, const double &in_vis_mu,
-        const double &in_beta, const int &elemtype, 
+        const double &in_beta, const FEType &elemtype, 
         const double &in_ct, const double &in_ctauc )
 : rho0( in_rho ), vis_mu( in_vis_mu ),
   alpha_f(tm_gAlpha->get_alpha_f()), alpha_m(tm_gAlpha->get_alpha_m()),
   gamma(tm_gAlpha->get_gamma()), beta(in_beta),
-  CI( (elemtype == 501 || elemtype == 601) ? 36.0 : 60.0 ),
+  CI( (elemtype == FEType::Tet4 || elemtype == FEType::Tet10 || elemtype == FEType::Tet10_v2) ? 36.0 : 60.0 ),
```
Comment (Mr-Chi-Ding) [inline examples/ns/src/PLocAssem_VMS_NS_GenAlpha.cpp:13]: This line is wrong.  

Here should be:  elemtype == FEType::Tet4 || elemtype == FEType::Hex8

**Comment by ju-liu** [comment]
需要注意 Tet10的编号是503，而Tet10_v2的编号是502.

**Comment by ju-liu** [comment]
FEType.hpp中的两个utility函数需要有个测试函数测试一下 @Mr-Chi-Ding

**Comment by Mr-Chi-Ding** [comment]
FEType.hpp 中的两个函数已通过测试。测试代码传到 examples/test/sys_test.cpp。
截屏2024-11-27 08 11 29

**Comment by Mr-Chi-Ding** [comment]
NS example has been tested. When using one process, there is no output difference between feature-enum-class and master branch. More detail can be seen in the attached file. No memory leak is detected.
屏幕截图 2024-11-29 103039
[test.pdf](https://github.com/user-attachments/files/17953680/test.pdf)

**Comment by Mr-Chi-Ding** [comment]
We have run both codes on two processes. There is no difference between two log files. See attaced.
[test_2p.pdf](https://github.com/user-attachments/files/17955781/test_2p.pdf)

**Comment by Mr-Chi-Ding** [comment]
We use 3 cpus to run all simulations.  Differences on cfd.log files have been identified for Tet4, Hex8, and Hex27. Details can be seen in the attached file.    [test_3.pdf](https://github.com/user-attachments/files/17959112/test_3.pdf)

**Comment by ju-liu** [comment]
> We use 3 cpus to run all simulations. Differences on cfd.log files have been identified for Tet4, Hex8, and Hex27. Details can be seen in the attached file. [test_3.pdf](https://github.com/user-attachments/files/17959112/test_3.pdf)

So it is the MPI.

**Comment by Mr-Chi-Ding** [comment]
linearPDE/transport has been tested. No memory leak is detected, and no difference on log files between enum-class and master branch when using 2 cpus. 
memory_check
[test_2p.pdf](https://github.com/user-attachments/files/18001340/test_2p.pdf)

**Comment by Mr-Chi-Ding** [comment]
Elastodynamics example has been tested. There is no memory leak and no differences on log files from enum-class and master branch.
elasto_memory_check
[test_elasto.pdf](https://github.com/user-attachments/files/18016756/test_elasto.pdf)

**Comment by Mr-Chi-Ding** [comment]
No memory leak has been detected from fsi codes on enum-class branch. There are no differences on FSI results between feature-enum-class and master branch. 
leak-memory-check
[test.pdf](https://github.com/user-attachments/files/18091393/test.pdf)

**Comment by Mr-Chi-Ding** [comment]
ALE codes has been tested. No memory leak has been deteced and no differences on log files from enum-class 
and master branches has been found.
leak_check_ALE
[test.pdf](https://github.com/user-attachments/files/18139322/test.pdf)

**Comment by Mr-Chi-Ding** [comment]
stress_recovery has been tested. No memory leak has been detected and no differences on log files have been found.
leak-check-stress
[test.pdf](https://github.com/user-attachments/files/18140633/test.pdf)

File: examples/ns/include/PLocAssem_VMS_NS_GenAlpha.hpp [hunk 2]
```diff
@@ -12,6 +12,8 @@
 #include "IPLocAssem.hpp"
 #include "TimeMethod_GenAlpha.hpp"
 #include "SymmTensor2_3D.hpp"
+#include "FEAElement.hpp"
+#include "FEType.hpp"
```
Comment (ju-liu) [inline examples/ns/include/PLocAssem_VMS_NS_GenAlpha.hpp:16]: 不需重复include了。FEType在FEAElement里已经包含了。

File: examples/ns/src/VTK_Writer_NS.cpp [hunk 3]
```diff
@@ -91,16 +91,16 @@ void VTK_Writer_NS::writeOutput(
         elemptr, dataVecs[1] );
 
     // Set mesh connectivity
```
Comment (ju-liu) [inline examples/ns/src/VTK_Writer_NS.cpp:93]: 我们是不是可以统一一下这里的调用方式？面谈

File: include/Vis_Tools.hpp [hunk 4]
```diff
@@ -142,20 +144,7 @@ namespace VIS_T
...
-  //   are indexed as:
-  //                ii_s + ii_t * seg_s
-  //   \para segs: number of sampling points in s direction
-  //   \para segt: number of sampling points in t direction
-  //   \para ptoffset: the offset for this element
-  //   \para gridData: the Grid object that stores the connectivity
-  // --------------------------------------------------------------
-  void setQuadelem( const int &segs, const int &segt,
-      const int &ptoffset, vtkUnstructuredGrid * gridData );
-
```
[the remark this answers is not in the document]
Comment (Mr-Chi-Ding) [reply include/Vis_Tools.hpp:158]: 这个函数确实没用过。

**Comment by Syj-0525** [comment]
[2024-12-17 enum-class-ns-test.pdf](https://github.com/user-attachments/files/18223982/2024-12-17.enum-class-ns-test.pdf)
已经完成tet10, hex8, hex27和主分支的对比测试，测试结果无论是输出的log文件，解文件，以及后处理文件都用diff对比后显示完全无差异

**Comment by SeaVegetable** [comment]
[Report_linearPDE_fsi_stress_recovery.pdf](https://github.com/user-attachments/files/18227966/Report_linearPDE_fsi_stress_recovery.pdf)
已完成linearPDE, fsi, stress_recovery的测试，在使用相同单元的情况下，该分支与主分支生成的log文件，可视化文件经diff对比后一致。

**Comment by Mr-Chi-Ding** [comment]
Update after Dec. 17, 2024:

(1) AGlobal_Mesh_Info.hpp:  remove “element_str”.

(2) Setquadelem in Vis_Tools.hpp.

(3) Rename vis_wss files in ALE, ns, and fsi examples.

(4) Use FE_T::to_FEType in Part_FEM.

(5) Remove two header files from ns/PLocAssem.

(6) Use in_elemtype in some BC classes like ElemBC_3D and NodalBC_inflow.

**Pull request merged**